src/core/io/writer/h5md_core.cpp

```diff
@@ -47,562 +52,340 @@ static void backup_file(const std::string &from, const std::string &to) {
-  H5Sclose(ds);
-  /* Extend the dataset for another timestep (extent = 1) */
-  for (int i = 0; i < rank; i++) {
-    dims[i] += change_extent[i];
-  }
-  H5Dset_extent(dataset.hid(), dims.data()); // extend all dims is collective
-}
-
-/* data is assumed to be three dimensional */
-template <typename T>
-void File::WriteDataset(T &data, const std::string &path,
-                        const std::vector<int> &change_extent, hsize_t *offset,
-                        hsize_t *count) {
-  ExtendDataset(path, change_extent);
-  auto &dataset = datasets[path];
-  hid_t ds = H5Dget_space(dataset.hid());
-  auto rank = static_cast<hsize_t>(H5Sget_simple_extent_ndims(ds));
-  std::vector<hsize_t> maxdims(rank);
-  for (int i = 0; i < rank; i++) {
-    maxdims[i] = H5S_UNLIMITED;
-  }
-  H5Sselect_hyperslab(ds, H5S_SELECT_SET, offset, nullptr, count, nullptr);
-  /* Create a temporary dataspace. */
-  hid_t ds_new = H5Screate_simple(rank, count, maxdims.data());
-  /* Finally write the data to the dataset. */
-  H5Dwrite(dataset.hid(), dataset.get_type(), ds_new, ds, H5P_DEFAULT,
-           data.origin());
-  H5Sclose(ds_new);
-  H5Sclose(ds);
-}
+void File::write(const ParticleRange &particles, double time, int step) {
+  write_box(box_geo, m_h5md_file, datasets["particles/atoms/box/edges/value"]);
+  write_connectivity(particles);
 
-void File::WriteScript(std::string const &filename) {
-  /* First get the number of lines of the script. */
-  hsize_t dims[1] = {1};
-  std::ifstream scriptfile(m_absolute_script_path.string());
-  /* Read the whole script into a buffer. */
-  scriptfile.seekg(0, std::ios::end);
-  auto filelen = scriptfile.tellg();
-  scriptfile.seekg(0);
-  std::vector<char> buffer;
-  buffer.reserve(static_cast<unsigned long>(filelen));
-  buffer.assign(std::istreambuf_iterator<char>(scriptfile),
-                std::istreambuf_iterator<char>());
-
-  hid_t dtype, space, dset, file_id;
-  file_id =
-      H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
-
-  dtype = H5Tcopy(H5T_C_S1);
-  H5Tset_size(dtype, static_cast<size_t>(filelen * sizeof(char)));
-
-  space = H5Screate_simple(1, dims, nullptr);
-  /* Create the dataset. */
-  hid_t link_crt_plist = H5Pcreate(H5P_LINK_CREATE);
-  H5Pset_create_intermediate_group(
-      link_crt_plist, static_cast<unsigned int>(
-                          true)); // Set flag for intermediate group creation
-  dset = H5Dcreate(file_id, "parameters/files/script", dtype, space,
-                   link_crt_plist, H5P_DEFAULT, H5P_DEFAULT);
-  /* Write data from buffer to dataset. */
-  H5Dwrite(dset, dtype, H5S_ALL, H5S_ALL, H5P_DEFAULT, buffer.data());
-  /* Clean up. */
-  H5Dclose(dset);
-  H5Sclose(space);
-  H5Tclose(dtype);
-  H5Fclose(file_id);
+  int const n_part_local = particles.size();
+  // calculate count and offset
+  int prefix = 0;
+  // calculate prefix for write of the current process
+  MPI_Exscan(&n_part_local, &prefix, 1, MPI_INT, MPI_SUM, m_comm);
```

Comment:
I think `exscan` is not available from boost.mpi, but it provides a macro for error checking, like so
```c++
BOOST_MPI_CHECK_RESULT(MPI_Exscan, (&n_part_local, &prefix, 1, MPI_INT, MPI_SUM, m_comm));
```